### PR TITLE
feat(modal): create initial implementation

### PR DESCRIPTION
**Pull request by wesleycho**
This is an initial implementation based off of https://github.com/angular/angular/blob/master/modules/angular2_material/src/components/dialog/dialog.ts (almost 1-1 port). This is still a work in progress, and needs tests too.

### CODE REVIEW

File: src/components/modal/modal.ts [hunk 1]
```diff
@@ -0,0 +1,196 @@
...
+  selector: 'Uib-backdrop',
+  host: {
+    '(click)': 'onClick()'
+  }
+})
+@View({template: '', encapsulation: ViewEncapsulation.None})
+class UibBackdrop {
+  modalRef: UibModalRef;
+
+  constructor(modalRef: UibModalRef, options: UibModalConfig) {
```
Comment (wesleycho) [inline src/components/modal/modal.ts]: I need to investigate whether the adding of options here is valid.

File: src/components/modal/modal.ts [hunk 2]
```diff
@@ -0,0 +1,196 @@
...
+  Injector,
+  View,
+  ViewEncapsulation
+} from 'angular2/core';
+
+import {ObservableWrapper, Promise, PromiseWrapper} from 'angular2/src/core/facade/async';
+import {isPresent, Type} from 'angular2/src/core/facade/lang';
+import {DOM} from 'angular2/src/core/dom/dom_adapter';
+import {MouseEvent, KeyboardEvent} from 'angular2/src/core/facade/browser';
+import {KeyCodes} from 'angular2_material/src/core/key_codes';
```
Comment (Foxandxss) [inline src/components/modal/modal.ts]: All this lines, perhaps we need to change that to `angular2/angular2` or similar. Seems that the import "source" changes depending on what angular2 package the end user have in their project or something like that. Need to research on that.

File: src/components/modal/modal.ts [hunk 3]
```diff
@@ -0,0 +1,196 @@
...
+
+@Injectable()
```
Comment (Foxandxss) [inline src/components/modal/modal.ts]: That technically is not needed anymore.

File: src/components/modal/modal.ts (hunk 2 above)
Comment (wesleycho) [inline src/components/modal/modal.ts]: Oh right, that is obviously bunk - I need to fix that heh.

File: src/components/modal/modal.ts [hunk 4]
```diff
@@ -0,0 +1,196 @@
...
+
+  documentKeypress(event: KeyboardEvent) {
+    if (event.keyCode === KeyCodes.ESCAPE) {
+      this.modalRef.close();
+    }
+  }
+}
+
+@Directive({
+  selector: 'Uib-modal-content'
```
Comment (Foxandxss) [inline src/components/modal/modal.ts]: bad uppercase.

File: src/components/modal/modal.ts [hunk 5]
```diff
@@ -0,0 +1,196 @@
...
+import {DOM} from 'angular2/src/core/dom/dom_adapter';
+import {MouseEvent, KeyboardEvent} from 'angular2/src/core/facade/browser';
+import {KeyCodes} from 'angular2_material/src/core/key_codes';
+
+@Injectable()
+export class UibModal {
+  componentLoader: DynamicComponentLoader;
+
+  constructor(loader: DynamicComponentLoader) {
+    this.componentLoader = loader;
```
Comment (Foxandxss) [inline src/components/modal/modal.ts]: Common convention I am seeing (but subjective I guess) is to put `private` on the constructor so we don't need line 29 and 26.

**Comment by wesleycho** [comment]
Closing as I want to re-think this completely - I want to make the modal observable-based, which will allow for possibilities such as canceling modal opening during the animation.

**Pull request closed without merging**